### PR TITLE
ipn/ipnlocal: handle key extensions after key has already expired

### LINKED ISSUE #2515: Issues with extend key/reauthentication
We have just reached 6 months since setting up tailscale, so most of our users are having to reauthenticate. 

I had to reauthenticate for a VM, extending the key did not let me connect. When I connected in via my physical machine Tailscale showed the greyed out icon (as if disconnected) and I right clicked and logged in, which got me working again.

One of my colleagues also had a key expire. In that case they did not have a way to access it, and I had to RDP from my VM in the office, into their machine via local IP. In that case tailscale appeared to be connected, and I just did a cycle of disconnect, reconnect. This seemed to resolve the connection issue, but the key is still showing as about to expire.

It's not clear from the instructions how/if you can reauthenticate a windows machine while connected to it via RDP over tailscale? For the time being I have turned off key expiry for users that do not have an alternative way of connecting to the machine.

### PR DESCRIPTION
**Pull request by maisem**
Fixes #2515
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -2358,7 +2360,14 @@ func (b *LocalBackend) enterState(newState ipn.State) {
...
 			systemd.Status("Stopped; run 'tailscale up' to log in")
 		}
-	case ipn.Starting, ipn.NeedsMachineAuth:
+	case ipn.Starting:
+		if oldState == ipn.NeedsLogin && !keyExpired && blocked {
+			// This means that key was extended **after** expiry,
+			// unblock engine updates.
+			b.blockEngineUpdates(false)
+		}
+		fallthrough
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:2369]: this isn't intuitive on its own (I understand how you got here from the old code, but people won't have that context later). e.g. How's Starting related to NeedsMachineAuth?

I'd rather you just copy/paste those two lines if we need them in this case.

Comment (maisem) [reply ipn/ipnlocal/local.go]: makes sense, copied the two lines.

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -2358,7 +2360,16 @@ func (b *LocalBackend) enterState(newState ipn.State) {
...
 		}
-	case ipn.Starting, ipn.NeedsMachineAuth:
+	case ipn.Starting:
+		if oldState == ipn.NeedsLogin && !keyExpired && blocked {
+			// This means that key was extended **after** expiry,
+			// unblock engine updates.
+			b.blockEngineUpdates(false)
+		}
+		b.authReconfig()
+		// Needed so that UpdateEndpoints can run
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:2370]: it's not obvious what line (prior or following) this comment is referring to, or why they make UpdateEndpoints run?

same on line 2374

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -2358,7 +2360,16 @@ func (b *LocalBackend) enterState(newState ipn.State) {
 		if authURL == "" {
 			systemd.Status("Stopped; run 'tailscale up' to log in")
 		}
-	case ipn.Starting, ipn.NeedsMachineAuth:
+	case ipn.Starting:
+		if oldState == ipn.NeedsLogin && !keyExpired && blocked {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:2364]: This all seems oddly specific.

Seems like there might be a bunch of ways to get stuck and we're addressing just one we saw.

Why is key expiry only sticking on NeedsLogin => Starting?

Why don't we unblock in LocalBackend.setClientStatus instead where it's already unblocked?

I mean, how do we know the `blocked` here being true means that it was blocked due to the key being expired?

Comment (maisem) [reply ipn/ipnlocal/local.go]: > Why don't we unblock in LocalBackend.setClientStatus instead where it's already unblocked?

That makes a lot more sense, how's this?

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -463,10 +464,18 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
 		return
 	}
 
+	keyExpired := st.NetMap != nil && !st.NetMap.Expiry.IsZero() && time.Until(st.NetMap.Expiry) <= 0
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:467]: time.Until reads a bit awkward to me at least here. Maybe `time.Now().After(st.NetMap.Expiry)` ?

**Review by bradfitz — COMMENTED** [review]
No updates to `ipn/ipnlocal/state_test.go` ?

File: ipn/ipnlocal/local.go (hunk 4 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Added an explicit check for NetMap and a comment to improve readability.

File: ipn/ipnlocal/local.go (hunk 2 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: I honestly have no idea what it does/means either, but I'm not changing that code now so leaving it as is.

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -463,10 +464,18 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
 		return
 	}
 
+	keyExpired := st.NetMap != nil && !st.NetMap.Expiry.IsZero() && time.Now().After(st.NetMap.Expiry)
 	b.mu.Lock()
 	wasBlocked := b.blocked
+	wasKeyExpired := b.keyExpired
+	b.keyExpired = keyExpired
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:471]: I think you only want to mutate b.keyExpired if st.NetMap != nil.  Otherwise it should be left how it was.

**Review by bradfitz — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 5 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: good point. makes the logic easier to follow too.

**Pull request merged**